### PR TITLE
[Improvement] Improve GraphAr spark writer performance and implement custom writer builder to bypass spark's write behavior

### PR DESCRIPTION
**Pull request by acezen**
## Proposed changes
This PR aims to improve GraphAr spark writer's performace by changes:
- Revise the edge DataFrame to edge chunk partition process:
   - Use sort and repartition by custom partitioner to do the edge chunk split job
- Implement custom writer builder for csv, parquet and orc to bypass the default write process of spark
- Implement a custom commit protocol class `GarCommitProtocol` to change the file strategy of spark write

## Types of changes

What types of changes does your code introduce to GraphAr?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/alibaba/GraphAr/blob/main/CONTRIBUTING.rst) doc
- [x] I have signed the CLA
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments
#68

### CODE REVIEW

File: spark/src/main/scala/com/alibaba/graphar/datasources/GarCommitProtocol.scala [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+import org.apache.spark.internal.Logging
+
+object GarCommitProtocol {
+  private def binarySearchPair(aggNums: Array[Int], key: Int): (Int, Int) = {
+    var low = 0
+    var high = aggNums.length - 1
+    var mid = 0
+    while (low <= high) {
+      mid = (high + low) / 2;
+      if (aggNums(mid) <= key && aggNums(mid + 1) > key) {
```
[the remark this answers is not in the document]
Comment (acezen) [reply spark/src/main/scala/com/alibaba/graphar/datasources/GarCommitProtocol.scala:36]: Actually, in our scenario, the key would always landing on index < aggNums.length-1, but we can revise the condition to `if (aggNums(mid) <= key && (mid == length - 1 || aggNums(mid + 1) > key))` for save.

File: spark/src/main/scala/com/alibaba/graphar/utils/Patitioner.scala [hunk 2]
```diff
@@ -46,3 +46,47 @@ class ChunkPartitioner(partitions: Int, chunk_size: Long) extends Partitioner {
...
+    }
+  }
+
+  private def binarySeach(key: Long): Int = {
+    var low = 0
+    var high = eidBeginOfVertexChunks.length - 1
+    var mid = 0
+    while (low <= high) {
+      mid = (high + low) / 2;
+      if (eidBeginOfVertexChunks(mid) <= key && eidBeginOfVertexChunks(mid + 1) > key) {
```
[the remark this answers is not in the document]
Comment (acezen) [reply spark/src/main/scala/com/alibaba/graphar/utils/Patitioner.scala:73]: Ditto.

File: spark/src/main/scala/com/alibaba/graphar/writer/EdgeWriter.scala [hunk 3]
```diff
@@ -31,75 +31,92 @@ import scala.collection.mutable.ArrayBuffer
...
-    val spark = chunkDf.sparkSession
-    import spark.implicits._
-    val df_schema = chunkDf.schema
-    val index = df_schema.fieldIndex(keyColumnName)
-    val df_rdd = chunkDf.rdd.map(row => (row(index).asInstanceOf[Long], row))
-
-    // generate global edge id for each record of dataframe
-    val parition_counts = df_rdd
+    val edgeSchema = edgeDf.schema
+    val colIndex = edgeSchema.fieldIndex(if (adjListType == AdjListType.ordered_by_source) GeneralParams.srcIndexCol else GeneralParams.dstIndexCol)
```
Comment (lixueclaire) [inline spark/src/main/scala/com/alibaba/graphar/writer/EdgeWriter.scala:41]: Is this function repartitionAndSort() only required for AdjListType.ordered_by_source & AdjListType.ordered_by_dest? It seems that it is called for all types.

File: spark/src/main/scala/com/alibaba/graphar/writer/EdgeWriter.scala [hunk 4]
```diff
@@ -31,75 +31,92 @@ import scala.collection.mutable.ArrayBuffer
...
+    val vertexChunkSize: Long = if (adjListType == AdjListType.ordered_by_source) edgeInfo.getSrc_chunk_size() else edgeInfo.getDst_chunk_size()
+    val edgeChunkSize: Long = edgeInfo.getChunk_size()
+    val vertexChunkNum: Int = ((vertexNumOfPrimaryVertexLabel + vertexChunkSize - 1) / vertexChunkSize).toInt  // ceil
+
+    // sort by primary key and generate continue edge id for edge records
+    val sortedDfRDD = edgeDf.sort(GeneralParams.srcIndexCol).rdd
```
Comment (lixueclaire) [inline spark/src/main/scala/com/alibaba/graphar/writer/EdgeWriter.scala:47]: "GeneralParams.srcIndexCol" -> "colIndex"

File: spark/src/main/scala/com/alibaba/graphar/writer/EdgeWriter.scala [hunk 5]
```diff
@@ -31,75 +31,83 @@ import scala.collection.mutable.ArrayBuffer
...
-    val partition_num = Math.ceil(chunkDf.count() / edgeChunkSize.toDouble).toInt
-    val partitioner = new ChunkPartitioner(partition_num, edgeChunkSize)
-    val chunks = rdd_with_eid.repartitionAndSortWithinPartitions(partitioner).values
-    spark.createDataFrame(chunks, df_schema)
+    // Construct partitioner for edge chunk
+    // get edge num of vertex chunks
+    val edgeNumOfVertexChunks = sortedDfRDD.mapPartitions(iterator => {
+      iterator.map(row => (row(colIndex).asInstanceOf[Long] / vertexChunkSize, 1))
+    }).reduceByKey(_ + _).collectAsMap()
+    val vertexChunkNum = edgeNumOfVertexChunks.size
```
[the remark this answers is not in the document]
Comment (acezen) [reply spark/src/main/scala/com/alibaba/graphar/writer/EdgeWriter.scala]: Resolved by passing a vertex num augment.

**Review by lixueclaire — APPROVED** [review]
LGTM.

**Pull request merged**